### PR TITLE
Typescript implementation

### PR DESCRIPTION
**Pull request by majo44**
This is PR which contains refactoring to typescript which:
- fallow patterns from redux-observable, so epic on input is consuming stream of events, and stream of state,
- where on state stream we have access to current state by `.value` property, and stream is filtered only to changed values of state (implementation taken directly from redux-observable)
- is strict static typed on the events, payload and state objects/streams 
- exposing `toEventObservable` which can be used for simple creation of stream of events
- exposing `toStateObservable` which can be used for simple creation of stream of state
- exposing `toEvent` which should be used on return from epic
- exposing `ofEvent` which only filter for known events, or even restrict epic to handle just subset of known events (input events)

Todo:
- [x] implementation
- [x] build of es5/commonjs 
- [x] build of es5/es6 module
- [x] linting
- [x] running jest tests
- [x] update readme
- [x] tsdoc
- [x] test coverage
- [x] size limit (looking for the places where we can optimize code)

### CODE REVIEW

**Comment by majo44** [comment]
Travis build is failed, and I'm not sure why, locally everything works fine, only difference is that I have newer yarn. :(

**Comment by distolma** [comment]
@majo44 Travis-CI build fails because `size-limit` tries to check `index.js` file, which doesn't exist before a build.

File: .travis.yml [hunk 1]
```diff
@@ -2,3 +2,5 @@ language: node_js
 cache: yarn
 node_js:
   - node
+before_install:
+  - npm install -g yarn
```
Comment (distolma) [inline .travis.yml]: Yarn is already installed in Travis build machine

Comment (majo44) [reply .travis.yml]: Yep I removed that

File: tsconfig.json [hunk 2]
```diff
@@ -1,22 +1,14 @@
 {
   "compilerOptions": {
     "moduleResolution": "node",
-    "target": "es6",
-    "module":"es2015",
-    "lib": ["es2016", "es2017", "dom"],
+    "target": "es5",
```
Comment (distolma) [inline tsconfig.json]: I think we can use es6 spec, or not?

**Comment by majo44** [comment]
@distolma thx, I missed that, so I will move the size-limit to build

File: .gitignore [hunk 3]
```diff
@@ -1,7 +1,17 @@
...
 node_modules/
-dist/
-coverage/
\ No newline at end of file
+coverage/
+
+# Release
+index.js
+index.d.ts
+index.js.map
```
Comment (distolma) [inline .gitignore]: `module` folder also neet to be ignored

File: package.json [hunk 4]
```diff
@@ -10,13 +10,16 @@
...
+  ],
   "license": "MIT",
   "scripts": {
     "start": "rollup -c rollup.config.ts -w",
-    "build": "tsc && tsc -p tsconfig.module.json",
+    "build": "tsc && tsc -p tsconfig.module.json && size-limit",
     "lint": "eslint index.ts",
-    "size": "size-limit",
-    "test": "jest --coverage && yarn lint && size-limit"
+    "test": "jest --coverage && yarn lint"
```
Comment (distolma) [inline package.json]: Better run `build` before size-limit, imho
```sh
jest --coverage && yarn lint && npm build && size-limit
```

File: package.json [hunk 5]
```diff
@@ -1,28 +1,37 @@
...
+  ],
   "license": "MIT",
   "scripts": {
     "start": "rollup -c rollup.config.ts -w",
-    "build": "rollup -c rollup.config.ts",
-    "lint": "eslint *.js",
-    "test": "jest --coverage && yarn lint && size-limit"
+    "build": "tsc && tsc -p tsconfig.module.json && size-limit",
+    "lint": "eslint index.ts",
+    "test": "jest --coverage && yarn lint"
```
Comment (distolma) [inline package.json]: Better run build before size-limit, imho

```sh
jest --coverage && yarn lint && npm build && size-limit
```

File: package.json [hunk 6]
```diff
@@ -37,18 +46,13 @@
...
     "storeon": "^1.0.0",
+    "ts-jest": "^25.2.1",
     "typescript": "^3.8.3"
   },
   "lint-staged": {
-    "src/**/*.ts": [
+    "index.ts": [
       "eslint --fix",
       "git add"
     ]
```
Comment (distolma) [inline package.json]: Should we check test files also?

File: package.json [hunk 7]
```diff
@@ -61,14 +65,21 @@
   "size-limit": [
     {
       "path": "index.js",
-      "limit": "226 B"
+      "limit": "600 B"
     }
```
Comment (distolma) [inline package.json]: Add also `module/index.js` file for size-limit check

File: tsconfig.json (hunk 2 above)
Comment (majo44) [reply tsconfig.json]: If we will use es6 then projects which use this lib will need to use typescript or babel loader for legacy browsers, I think for now is better to stay with es5, is simple to change in future.

File: .travis.yml (hunk 1 above)
Comment (majo44) [reply .travis.yml]: done

File: .gitignore (hunk 3 above)
Comment (majo44) [reply .gitignore]: done

File: package.json (hunk 4 above)
Comment (majo44) [reply package.json]: I changed the scripts, so now `build` is main script which orchestrate all other

File: package.json (hunk 5 above)
Comment (majo44) [reply package.json]: I changed the scripts, so now build is main script which orchestrate all other

File: package.json (hunk 6 above)
Comment (majo44) [reply package.json]: ok

File: tsconfig.json (hunk 2 above)
Comment (distolma) [reply tsconfig.json]: `storeon` itself use [es6 spec](https://github.com/storeon/storeon/commit/82cecf51c9fd38d1e62875dc4df5087899b2743a), so user anyway needs to use typescript or babel loader for legacy browsers

Comment (majo44) [reply tsconfig.json]: You right, ok lets go to es6.

**Comment by distolma** [comment]
Great job!

**Pull request merged**